xla/tests/collective_ops_e2e_test.cc

```diff
@@ -1340,6 +1341,295 @@ TEST_F(CollectiveOpsTestE2E, NoAsyncCollectives) {
+            upper[root_sharded_dims[m]] +=
+                root_dims_per_shard[root_sharded_dims[m]];
+            break;
+          } else {
+            upper[root_sharded_dims[m]] =
+                root_dims_per_shard[root_sharded_dims[m]];
+          }
+        }
+        std::transform(upper.begin(), upper.end(), root_dims_per_shard.begin(),
+                       lower.begin(), std::minus<int64_t>());
+      }
+    } else {
+      EXPECT_TRUE(
+          LiteralTestUtil::Near(ref_results[0], results[0], error_spec));
+    }
+  }
+
+  void EvaluateShardedDims(std::vector<int64_t>& dims_per_shard,
+                           std::vector<int64_t>& sharded_dims,
+                           const HloSharding& sharding) {
+    if (!sharding.IsReplicated()) {
+      for (int k = 0; k < sharding.tile_assignment().num_dimensions(); ++k) {
+        if (sharding.tile_assignment().dim(k) > 1) {
+          dims_per_shard[k] /= sharding.tile_assignment().dim(k);
+          sharded_dims.push_back(k);
+        }
+      }
+    }
+  }
+};
+
+TEST_F(CollectiveOpsTestE2EShardedUnsharded, DotBatchAndBatch) {
+  const std::string hlo_text = R"(
+HloModule module, entry_computation_layout={(f32[4,16,8]{2,1,0}, f32[4,4,8]{2,1,0})->f32[4,16,4]{2,1,0}}, num_partitions=2
+
+ENTRY entry {
+  lhs = f32[4,16,8]{2,1,0} parameter(0), sharding={devices=[2,1,1]<=[2]}
+  rhs = f32[4,4,8]{2,1,0} parameter(1), sharding={devices=[2,1,1]<=[2]}
+  ROOT dot = f32[4,16,4]{2,1,0} dot(lhs, rhs), lhs_batch_dims={0}, rhs_batch_dims={0}, lhs_contracting_dims={2}, rhs_contracting_dims={2}, sharding={devices=[1,2,1]<=[2]}
+})";
+  CollectiveOpsCompareShardedUnsharded(hlo_text);
+}
+
+TEST_F(CollectiveOpsTestE2EShardedUnsharded, DotBatchAndNonContracting) {
+  const std::string hlo_text = R"(
+HloModule module, entry_computation_layout={(f32[4,16,8]{2,1,0}, f32[4,4,8]{2,1,0})->f32[4,16,4]{2,1,0}}, num_partitions=2
+
+ENTRY entry {
+  lhs = f32[4,16,8]{2,1,0} parameter(0), sharding={devices=[2,1,1]<=[2]}
+  rhs = f32[4,4,8]{2,1,0} parameter(1), sharding={devices=[1,2,1]<=[2]}
+  ROOT dot = f32[4,16,4]{2,1,0} dot(lhs, rhs), lhs_batch_dims={0}, rhs_batch_dims={0}, lhs_contracting_dims={2}, rhs_contracting_dims={2}, sharding={devices=[2,1,1]<=[2]}
+})";
+  CollectiveOpsCompareShardedUnsharded(hlo_text);
+}
+
+TEST_F(CollectiveOpsTestE2EShardedUnsharded, DotContractingAndContracting) {
+  const std::string hlo_text = R"(
+HloModule module, entry_computation_layout={(f32[16,8]{1,0}, f32[4,8]{1,0})->f32[16,4]{1,0}}, num_partitions=2
+
+ENTRY entry {
+  lhs = f32[16,8]{1,0} parameter(0), sharding={devices=[1,2]<=[2]}
+  rhs = f32[4,8]{1,0} parameter(1), sharding={devices=[1,2]<=[2]}
+  ROOT dot = f32[16,4]{1,0} dot(lhs, rhs), lhs_contracting_dims={1}, rhs_contracting_dims={1}, sharding={devices=[2,1]<=[2]}
+})";
+  CollectiveOpsCompareShardedUnsharded(hlo_text);
+}
+
+TEST_F(CollectiveOpsTestE2EShardedUnsharded, DotNonContractingAndContracting) {
```

Comment:
This test crashed internally (running on 2 GPUs). Any quick guesses on why?
```
[ RUN      ] CollectiveOpsTestE2EShardedUnsharded.DotNonContractingAndContracting
*** SIGILL (@0x558221a2da23), received by PID 2659 (TID 2659) on cpu 8; stack trace: ***
xla::(anonymous namespace)::CollectiveOpsTestE2EShardedUnsharded::CollectiveOpsCompareShardedUnsharded()
FailureSignalHandler()
xla::(anonymous namespace)::CollectiveOpsTestE2EShardedUnsharded::CollectiveOpsCompareShardedUnsharded()
xla::(anonymous namespace)::CollectiveOpsTestE2EShardedUnsharded_DotNonContractingAndContracting_Test::TestBody()
testing::Test::Run()
testing::TestInfo::Run()
testing::TestSuite::Run()
testing::internal::UnitTestImpl::RunAllTests()
testing::UnitTest::Run()
main
...
```

Reply:
@penpornk all tests pass on my side.

Reply:
@penpornk can you try again with [f7bc551](https://github.com/openxla/xla/pull/29511/commits/f7bc551a60d741a52add7734dcaccad061d376ae)?